src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala

```diff
@@ -70,19 +81,133 @@ class GoogleDataprocDAO(protected val dataprocConfig: DataprocConfig)(implicit v
     }
   }
 
-
-  def deleteCluster(googleProject: String, clusterName: String)(implicit executionContext: ExecutionContext): Future[Unit] = {
+  override def deleteCluster(googleProject: String, clusterName: String)(implicit executionContext: ExecutionContext): Future[Unit] = {
     Future {
       //currently, the bucketPath of the clusterRequest are not used - it will be used later as a place to store notebooks and results
-      val request = dataproc.projects().regions().clusters().delete(googleProject, dataprocConfig.dataprocDefaultZone, clusterName)
+      val request = dataproc.projects().regions().clusters().delete(googleProject, dataprocConfig.dataprocDefaultRegion, clusterName)
       try {
         executeGoogleRequest(request)
       } catch {
-        case e:GoogleJsonResponseException =>
-          if(e.getStatusCode!=404)
+        case e: GoogleJsonResponseException =>
+          if (e.getStatusCode != 404)
             throw CallToGoogleApiFailedException(googleProject, clusterName, e.getStatusCode, e.getDetails.getMessage)
       }
     }
   }
 
+  override def getClusterStatus(googleProject: GoogleProject, clusterName: String)(implicit executionContext: ExecutionContext): Future[ClusterStatus] = {
+    getCluster(googleProject, clusterName).map { cluster =>
+      ClusterStatus.withNameIgnoreCase(cluster.getStatus.getState)
+    }
+  }
+
+  override def getClusterMasterInstanceIp(googleProject: GoogleProject, clusterName: String)(implicit executionContext: ExecutionContext): Future[Option[String]] = {
+    val ipOpt = for {
+      cluster <- OptionT.liftF[Future, Cluster] { getCluster(googleProject, clusterName) }
+      masterInstanceName <- OptionT.fromOption { getMasterInstanceName(cluster) }
+      masterInstanceZone <- OptionT.fromOption { getZone(cluster) }
+      masterInstance <- OptionT.liftF[Future, Instance] { getInstance(googleProject, masterInstanceZone, masterInstanceName) }
+      masterInstanceIp <- OptionT.fromOption { getInstanceIP(masterInstance) }
+    } yield masterInstanceIp
+
+    ipOpt.value
+  }
+
+  override def getClusterErrorDetails(operationName: String)(implicit executionContext: ExecutionContext): Future[Option[ClusterErrorDetails]] = {
+    val errorOpt = for {
+      operation <- OptionT.liftF[Future, Operation] { getOperation(operationName) }
+      done <- OptionT.pure { operation.getDone } if done == true
+      error <- OptionT.pure { operation.getError }
+      code <- OptionT.pure { error.getCode }
+    } yield ClusterErrorDetails(code, Option(error.getMessage))
+
+    errorOpt.value
+  }
+
+  /**
+    * Gets a dataproc Cluster from the API.
+    */
+  private def getCluster(googleProject: GoogleProject, clusterName: String)(implicit executionContext: ExecutionContext): Future[Cluster] = {
+    val request = dataproc.projects().regions().clusters().get(googleProject, dataprocConfig.dataprocDefaultRegion, clusterName)
+    Future {
```

Comment:
Out of curiosity - is it better to have `Future{...}` wrapped just around `executeGoogleRequest` or the entire body of the function, the way it is in the `build(...)` function in this file. Does it not matter or should we choose one way and use it consistently throughout?

Reply:
I like just wrapping the blocking calls inside Future. Creating the Google request is just invoking constructors so it doesn't really need to be asynchronous.

Actually I'm going to propose we make all Google calls through this method (I haven't pushed it yet):

```
  private def executeGoogleRequestAsync[A](googleProject: GoogleProject, clusterName: String)(request: AbstractGoogleClientRequest[A])(implicit executionContext: ExecutionContext): Future[A] = {
    Future {
     executeGoogleRequest(request)
    } recover {
      case e: GoogleJsonResponseException =>
        throw CallToGoogleApiFailedException(googleProject, clusterName, e.getStatusCode, e.getDetails.getMessage)
    }
  }
```

Reply:
That's a great idea!